File: frontend/src/types/game.ts
export interface ScanRequest {
	type: "scan"
	input: number
}

export interface Coord {
	x: number
	y: number
}

export interface SuggestRequest {
	type: "suggest"
	coords: Coord[]
}

export interface StartRequest {
	type: "start"
	settings: Settings
}


export type ClientMessage =
	ScanRequest
	| SuggestRequest
	| StartRequest


export interface ScanResponse {
	type: "scan"
	input: number
	output: number
}

export interface SuggestResponse {
	type: "suggest"
	success: boolean
}

export interface Settings {
	sizeX: number
	sizeY: number
	atomsCount: number
}

export type StartResponse = StartRequest

export interface ErrorResponse {
	type: "error"
	error: "TODO"
}

export type ServerMessage =
	ErrorResponse
	| ScanResponse
	| SuggestResponse
	| StartResponse

File: backend/app/model/game.py
from typing import Literal, Union

from pydantic import Field

from app.model.common import DefaultModel


class GameSettings(DefaultModel):
    size_x: int
    size_y: int
    atoms_count: int


class GameCreateRequest(DefaultModel):
    settings: GameSettings


class ScanRequest(DefaultModel):
    type: Literal["scan"]
    input: int


class Coord(DefaultModel):
    x: int
    y: int


class SuggestRequest(DefaultModel):
    type: Literal["suggest"]
    coords: list[Coord]


class Start(DefaultModel):
    type: Literal["start"]
    settings: GameSettings


class ClientMessage(DefaultModel):
    message: Union[ScanRequest, SuggestRequest, Start] = Field(
        ..., discriminator="type"
    )


class ScanResponse(DefaultModel):
    type: Literal["scan"]
    input: int
    output: int


class SuggestResponse(DefaultModel):
    type: Literal["suggest"]
    success: bool


class ErrorResponse(DefaultModel):
    type: Literal["error"]
    error: str
